src/Microsoft.ML.Core/Utilities/ThreadUtils.cs

```diff
@@ -4,44 +4,154 @@
+
+            private static Task Queue(Delegate threadStart, object state)
             {
-                IsBackground = false
-            };
+                // Create the TaskCompletionSource used to represent this work.
+                // Call sites only care about completion, not about the distinction between
+                // success and failure and do not expect exceptions to be propagated in this manner,
+                // so only SetResult is used.
+                var tcs = new TaskCompletionSource<bool>(TaskContinuationOptions.RunContinuationsAsynchronously);
+
+                // Queue the work for a thread to pick up.  If no thread is immediately available, create one.
+                if (!Enqueue((threadStart, state, tcs)))
+                {
+                    CreateThread();
+                }
+
+                // Return the task.
+                return tcs.Task;
+
+                void CreateThread()
+                {
+                    // Create a new background thread to run the work.
+                    var t = new Thread(() =>
+                    {
+                        // Repeatedly get the next item and invoke it, setting its TCS when we're done.
+                        // This will wait for up to the idle time before giving up and exiting.
+                        while (TryDequeue(out (Delegate, object, TaskCompletionSource<bool>) item))
+                        {
+                            try
+                            {
+                                if (item.Item1 is Action<object> pts)
+                                {
+                                    pts(item.Item2);
+                                }
+                                else
+                                {
+                                    ((Action)item.Item1)();
+                                }
+                            }
+                            finally
+                            {
+                                item.Item3.SetResult(true);
+                            }
+                        }
+                    });
+                    t.IsBackground = true;
+                    t.Start();
+                }
+
+                bool Enqueue((Delegate, object, TaskCompletionSource<bool>) item)
+                {
+                    // Enqueue the work.  If there are currently fewer threads waiting
+                    // for work than there are work items in the queue, create another
+                    // thread.  This is a heuristic, in that we might end up creating
+                    // more threads than are truly needed, but this whole type is being
+                    // used to replace a previous solution where every work item created
+                    // its own thread, so this is an improvement regardless of any
+                    // such inefficiencies.
+                    lock (_queue)
```

Comment:
I'm confused. While a thread is waiting for work, it holds the lock on `_queue`. How can Enqueue take the lock such that it can provide work?

Reply:
Oh never mind, I forgot that Monitor.Wait releases the lock.